M2/Macaulay2/packages/SCMAlgebras.m2

```diff
@@ -0,0 +1,394 @@
+      filterIdeal(J,5)
+  SeeAlso
+    minDim
+    isSCM
+///);
+-------------------------------------------------------------------------
+filterIdeal = method(TypicalValue=>Ideal);
+-------------------------------------------------------------------------
+filterIdeal(Ideal,ZZ) := Ideal => (I,i) -> (
+    S:=ring I;
+    d:=dim I;
+    d0:=minDim I;
+    if i>-2 and i<d0 then (
+        Ii:=I;
+    ) else if i>=d0 and i<d then ( 
+        Q:=decompose I;
+        L:={};
+        for k from 0 to #Q-1 do (
+            Qk:=Q#k;
+            Pk:=radical Qk;
+            dk:=dim Pk;
+            L=L|{{dk,Qk}};
+        );
+        Pi:={};
+        for k from 0 to #L-1 do (
+            if (L#k)#0 > i then (
+                Pi=Pi|{(L#k)#1};
+            );
+        );
+        Ii=intersect(Pi);
+    ) else if i==d then (
+        Ii=ideal(S^1);
+    ) else error("Expected an integer greater than -2 and at most " | toString(d) | ".");
+    return Ii
+)
+--=======================================================================
+
+
+--=======================================================================
+-- checks whether an ideal/module is sequentially Cohen-Macaulay
+-- -- for a module the Schenzel's characterization is used by default
+-- -- for an ideal the Goodarzi's characterization is used by default
+--=======================================================================
+MyDoc=concatenate(MyDoc,///
+Node
+  Key
+    isSCM
+    (isSCM,Module)
+    (isSCM,Ideal)
+  Headline
+    checks whether a module or an ideal is sequentially Cohen-Macaulay
+  Usage
+    isSCM M
+    isSCM I
+  Inputs
+    M:Module
+      a finitely generated graded $S$-module, $S=K[x_1,\ldots,x_n]$, with $K$ a field or an ideal $I\subset S$
+  Outputs
+    b:Boolean
+      whether the module $M$ or the graded algebra $S/I$ is sequentially Cohen-Macaulay
+  Description
+    Text
+      Given a finitely generated graded $S$-module $M$, this method checks if the $i$th module of deficiency of $M$, $\omega^{i}(M)$, if non-zero, is Cohen-Macaulay of dimension $i$.
+      For a homogeneous ideal $I\subset S$, the function checks if $\mathrm{depth} S/{I^{<i>}} \geq i+1$, where $I^{<i>}$ is the $i$th filter ideal.
+    Example
+      S=QQ[x_1..x_5];
+      M=coker matrix{{x_1*x_2,x_3*x_4,0,0},{0,x_1*x_5,x_2*x_4,0}};
+      isSCM M
+    Example
+      S = QQ[x_1..x_10,y_1..y_10];
+      E = {{1,2},{1,3},{1,4},{1,5},{1,6},{1,7},{1,8},{1,9},{1,10},{6,7},{8,9},{8,10},{9,10}};
+      J=ideal(for e in E list x_(e#0)*y_(e#1)-x_(e#1)*y_(e#0));
+      isSCM J
+///);
+-------------------------------------------------------------------------
+isSCM = method(TypicalValue=>Boolean);
+-------------------------------------------------------------------------
+isSCM(Module) := M -> (
+    d:=dim M;
+    for i from 0 to d-1 do ( 
+        Oi:=deficiencyModule(M,i);
+        if  Oi != 0 then (
+            if (isCM(Oi) == false) or (dim(Oi)!=i) then (
```

Comment:
From the style guide:

> Don't test a Boolean value for equality with true or false, because x===true is equivalent to x and x===false is equivalent to not x.

Here, you can use `if not isCM Oi or ...`.